### ISSUE TITLE
Custom code # could not be added to abc.php due to changes in....

### ISSUE THREAD

**Issue by duetnet** [vdm-io/Joomla-Component-Builder#731]
### Steps to reproduce the issue
This problem is very hard to pin down as it has occurred at random without me understanding the pattern, but I will try to explain what happened this time and what my clues to the issue are.
I have existing replacement code in admin view models.
JCB has compiled the component repeatedly the past few days with NO problem
Today I added 3 new site view templates.
Compiled the component and warning messages for all my REPLACEMENT custom code popped up.
PLEASE NOTE: I always compile with DEBUG LINE NUMBERS = ON whilst in development.
### Expected result
No problems with compile.
### Actual result
The compiler displayed warning messages displayed for 20 replacement codes, e.g.
- Custom code #15 could not be added to admin/models/job.php please review the file after install at line 977 and reposition the code, remove the comments and recompile to fix the issue. The issue could be due to a change to lines below the custom code.

### System information (as much as possible)
PHP Built On 	:	Windows NT DIANA 10.0 build 19041 (Windows 10) AMD64
Database Type 	:	mysql
Database 		:	5.5.5-10.4.17-MariaDB
PHP Version 	:	8.0.1
Web Server 	:	Apache/2.4.46 (Win64) OpenSSL/1.1.1h PHP/8.0.1
Joomla 		:	Joomla! 3.9.25 Stable [ Amani ] 02-March-2021 15:00 GMT
User Agent 	:	Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:87.0) Gecko/20100101 Firefox/87.0 
### Additional comments
What I have determined is the following:

- The only difference between the PRE and POST compiled versions, prior to the REPLACEMENT and after the REPLACEMENT code is the comment line number for every "// [Interpretation XXXX]" has changed consistently in the POST version. 
- No other line of code has changed in the file above or below the original REPLACEMENT point, except for the fact that  the post compile version has comment out the original REPLACEMENT and added back the original JCB code.
- Every POST compile comment line as new INTERPRETATION line number which is 10 digits higher than the PRE complie comment line. 

e.g.

_**PRE COMPILE VERSION**_
	public function getForm($data = array(), $loadData = true, $options = array('control' => 'jform'))
	{
		// set load data option
		$options['load_data'] = $loadData;
		// **[Interpretation 19566]** check if xpath was set in options
		$xpath = false;
		if (isset($options['xpath']))

**POST COMPILE VERSION**
	public function getForm($data = array(), $loadData = true, $options = array('control' => 'jform'))
	{
		// set load data option
		$options['load_data'] = $loadData;
		// **[Interpretation 19576]** check if xpath was set in options
		$xpath = false;
		if (isset($options['xpath']))

So, at the point of replacement the PRE and POST compile code is as follows:

_**PRE COMPILE VERSION**_

/***[REPLACED$$$$]***//*4*/
			$data['published'] = 1;  
/***[/REPLACED$$$$]***/			 
		}
		// [Interpretation **9878**] Automatic handling of alias for empty fields
		if (in_array($input->get('task'), array('apply', 'save', 'save2new')) && (int) $input->get('id') == 0)
		{
			if ($data['alias'] == null || empty($data['alias']))

_**POST COMPILE VERSION**_

/***[REPLACED$$$$]***//*4*/
// 			$data['published'] = 1;
// 
/***[/REPLACED$$$$]***/
			$data['published'] = 0;
		}
		// [Interpretation **9888**] Automatic handling of alias for empty fields
		if (in_array($input->get('task'), array('apply', 'save', 'save2new')) && (int) $input->get('id') == 0)
		{
			if ($data['alias'] == null || empty($data['alias']))

I deduce, therefore, that the HASH code generated by JCB for the REPLACEMENT has taken the INTERPRETATION number into account and this has caused the problem.

If I compile the component WITHOUT DEBUG LINE NUMBERS, the problem still arises because the original PRE COMPILE version still had the DEBUG line numbers and so the HASH code was generated whilst the INTERPRETATION numbers were present.

In conclusion, my only safe solution going forward is to ALWAYS compile WITHOUT debug line numbers. 
_**ONE PROBLEM HERE** : This compiler setting refers to GLOBAL PARAMETER for DEBUG LINE NUMBERS, but this setting does not seem to be listed on any of the JCB CONFIGURATION settings TABS_

The nagging question in my mind is, why have the INTERPRETATION numbers changed when I have not made any changes other than adding a few templates to a site view?

**Comment by Llewellynvdm** [comment]
The custom code feature is powerful, but has its drawbacks... this is why I made [this video](https://www.youtube.com/watch?v=QGMURJI_Mds&list=PLQRGFI8XZ_wuzAQELTUUIDdqKFmLP9lSl&index=11).

Most of the time when we consistently change code that JCB generates... why not contribute those improvements to the JCB project. Let us try and automate the change... this is my workflow. Soon as I see I need a change in the produced code... I look if there is a easy JCB way. Then if not I start working on a new feature to improve JCB to do that. So yes I am still moveing the tool to serve my own needs... it is why I build it. 

I am willing to start letting others enjoy the same freedom. I am even look at way they can do that without touching the core.

# Then if we use custom code

To use line numbers or not to use line numbers?

- It is a debugging feature for those working on the core compiler.
- It can change at any time as we improve the compiler
- They do effect the placement of code at this time (we could remove that... but...)
- The should not be in production code

These are some if the facts you should look at when you use line numbers in your code.

Now the global setting of DEBUG LINE NUMBERS are set per/component. So in you component you can set the global for that component... I have not though of adding it to the Global Config... but that could also be an idea. For now it is just in the component.

![image](https://user-images.githubusercontent.com/5607939/115560323-c5b94b00-a2b4-11eb-8ac5-4eca22685f41.png)

These three setting are the global option per/component. Maybe global was not the right name... but ya.

**Comment by duetnet** [comment]
Thank you for feedback, and, yes I understand. I suppose I have been using the debugging feature to try and keep track of what is happening the JCB source whilst I am developing. But I will now switch it off for future development as it has become a problem in migration of systems in development.

So sorry, I missed the 'default' settings on the component, very silly of me.
